Add geostrophic convergence test (Williamson Test 2)

xylar opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

This merge adds the `geostrophic` convergence test case from [Williamson et al. 1992](<https://doi.org/10.1016/S0021-9991(05)80016-6>).  The test case starts out in geostrophic balance and ideally remains in steady state.

The test case has similarities to `cosine_bell` so efforts have been made to share code, primarily the `forward` step, between the two.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.md`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://e3sm-project.github.io/polaris/main/developers_guide/building_docs.html) and changes look as expected
* [x] `Testing` comment in the PR documents testing used to verify the changes
* [x] New tests have been added to a test suite

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar commented:
This is based off of #119 and will be rebased after that gets merged.

xylar commented:
Thanks to @lconlon for starting out this test case with me in the Polaris Hackathon in May!

xylar commented:
## Testing

I have run all 4 geostrophic test cases on Chrysalis with Intel and OpenMPI.  They ran successfully and produced the expected plots.

The order of convergence for water-column thickness for the QU grids is really terrible -- 0.4.  So that's discoraging for MPAS-Ocean but not an indication of a problem with the tests.  The Icos meshes do much better -- about 1.6.

QU:
![convergence_h](https://github.com/E3SM-Project/polaris/assets/4179064/3f561874-c824-439b-9732-b416b54f12f4)
![convergence_vel](https://github.com/E3SM-Project/polaris/assets/4179064/3e922540-454f-433a-b14a-8327a0274c39)

Icos:
![convergence_h](https://github.com/E3SM-Project/polaris/assets/4179064/dba947b9-70a7-42c6-9da9-857d6e9a11a9)
![convergence_vel](https://github.com/E3SM-Project/polaris/assets/4179064/a894124f-db7d-4120-a94c-1d3a681edf3e)

xylar commented:
Here are some plots of water-column thickness, u, v and normal velocity (for the icos 240km mesh):

![final_h](https://github.com/E3SM-Project/polaris/assets/4179064/936c0374-4d14-42df-809d-96aa57a015d4)
![final_u](https://github.com/E3SM-Project/polaris/assets/4179064/2a0a4f30-15f4-4c6c-80f1-507954e32421)
![final_v](https://github.com/E3SM-Project/polaris/assets/4179064/06cdb4b7-5d84-45c8-85a6-b56ef9588505)
![final_norm_vel](https://github.com/E3SM-Project/polaris/assets/4179064/be2b0be6-a3db-43ba-a227-5d93ff0c1951)

xylar commented:
Here are diff plots of water-column thickness for the QU and Icos meshes with increasing resolution.

# QU

![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/97154233-11a4-41a0-aa7e-a2ed24a67c93)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/d40c817d-eb34-40cb-bb4e-05cf665b3c2c)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/9d06fbf8-8634-4329-b996-b7f1ec5095e8)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/740619c2-e9cd-444d-88c6-f49250e2f64e)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/90edc97f-5ea5-48fa-9316-94e26ae740e6)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/db2c3e1c-1223-421f-8005-70963ffce8c7)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/aaf8d0ae-17ea-4c7c-9438-c41c42f93d57)

# Icos

![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/5b217561-6c6b-44e5-ab50-aceec20821b6)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/20ac9acd-5fdb-4f98-83d5-2f676cb92ed3)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/9cc856c2-7f77-4227-9d63-26543a9a0e93)
![diff_h](https://github.com/E3SM-Project/polaris/assets/4179064/5fc10881-f955-4a4f-95b4-835bc6e7b612)

xylar commented:
I need to update this to use #126, so it will likely make sense to wait on reviewing this until after that PR has gone in.

xylar commented:
@cbegeman, just an update that this is coming along and now uses the shared analysis, but I need to test it.

xylar commented on polaris/ocean/convergence/spherical/analysis.py:
```diff
@@ -352,6 +348,39 @@ def exact_solution(self, mesh_name, field_name, time, zidx=None):
 
         return ds_init[field_name]
 
+    def get_output_field(self, mesh_name, field_name, time, zidx=None):
+        """
+        Get the model output field at the given time and z index
+
+        Parameters
+        ----------
+        mesh_name : str
+            The mesh name which is the prefix for the output file
+
+        field_name : str
+            The name of the variable of which we evaluate convergence
+
+        time : float
+            The time at which to evaluate the exact solution in seconds
+
+        zidx : int, optional
+            The z-index for the vertical level to take the field from
+
+        Returns
+        -------
+        field_mpas : xarray.DataArray
+            model output field
+        """
+        ds_out = xr.open_dataset(f'{mesh_name}_output.nc')
+
+        tidx = _time_index_from_xtime(ds_out.xtime.values, time)
+        ds_out = ds_out.isel(Time=tidx)
+
+        field_mpas = ds_out[field_name]
+        if zidx is not None:
+            field_mpas = field_mpas.isel(nVertLevels=zidx)
+        return field_mpas
```
@cbegeman, I need to make this a method similar to `exact_solution()` because I need a way to calculate the water-column thickness (which is not an output variable).
1. Would you prefer a different name for the method?
2. Are you okay with this proposed approach?

cbegeman replied:
@xylar That looks good to me. My original thought was just that folks would add whatever variables they need to the output netcdf file but it's probably better this way so we don't modify those files.

xylar replied:
Thanks for the feedback!

Yeah, I don't think modifying files that a step doesn't "own" is a good idea and it's not possible to modify the files in the forward step itself because it's a "command-line" step that just calls the model (no python code allowed).

cbegeman commented:
@xylar Just an FYI that I'll be posting a PR soon with some additional clean-up of the convergence analysis step. We might want to get this reviewed and merged first so that you don't have to do another rebase.

cbegeman commented on polaris/ocean/suites/geostropic.txt:
```diff
@@ -0,0 +1,2 @@
+ocean/spherical/icos/geostrophic
+ocean/spherical/qu/geostrophic
```
@xylar This file name is misspelled

xylar replied:
It sure is!  Should be fixed now.

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+
+        component = self.component
+
+        analysis_dependencies: Dict[str, Dict[str, Step]] = (
```
```suggestion
        analysis_dependencies: Dict[str, Dict[float, Step]] = (
```

`resolutions` appears to be a list of floats

xylar replied:
Excellent point.  Does cosine bell have the same problem?

cbegeman replied:
Oddly, `cosine_bell` doesn't give an error but `inertial_gravity_wave` did when I implemented it there. However, this fix works for both.

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+            self.add_step(base_mesh_step, symlink=f'base_mesh/{mesh_name}')
+            analysis_dependencies['mesh'][resolution] = base_mesh_step
+
+            cos_bell_dir = f'spherical/{prefix}/geostrophic'
```
```suggestion
            case_dir = f'spherical/{prefix}/geostrophic'
```

xylar replied:
Shoot, this must have been a sloppy copy-paste job after the last rebase.

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+            cos_bell_dir = f'spherical/{prefix}/geostrophic'
+
+            name = f'{prefix}_init_{mesh_name}'
+            subdir = f'{cos_bell_dir}/init/{mesh_name}'
```
```suggestion
            subdir = f'{case_dir}/init/{mesh_name}'
```

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+            analysis_dependencies['init'][resolution] = init_step
+
+            name = f'{prefix}_forward_{mesh_name}'
+            subdir = f'{cos_bell_dir}/forward/{mesh_name}'
```
```suggestion
            subdir = f'{case_dir}/forward/{mesh_name}'
```

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+            analysis_dependencies['forward'][resolution] = forward_step
+
+            if self.include_viz:
+                with_viz_dir = f'spherical/{prefix}/geostrophic/with_viz'
```
```suggestion
                with_viz_dir = f'{case_dir}/with_viz'
```

cbegeman commented on polaris/ocean/tasks/geostrophic/__init__.py:
```diff
@@ -0,0 +1,206 @@
+                step.set_shared_config(config, link=config_filename)
+                self.add_step(step)
+
+        subdir = f'spherical/{prefix}/geostrophic/analysis'
```
```suggestion
        subdir = f'{case_dir}/analysis'
```

cbegeman commented on polaris/ocean/tasks/geostrophic/analysis.py:
```diff
@@ -0,0 +1,136 @@
+            ssh = super().get_output_field(mesh_name=mesh_name,
+                                           field_name='ssh',
+                                           time=time, zidx=None)
+            h = ssh + bottom_depth
```
In other cases with constant bottom_depth we just evaluate convergence with ssh. Are you just using wct for clarity or is there a reason I'm missing?

xylar replied:
I had a lot of trouble initially and only realize after some trial an error that it was because I was using a `z-level` rather than a `z-star` coordinate.  With `z-star`, you can have the bathymetry at zero, but I don't really love that approach.  It feels to me relatively important to use water-column thickness if that is the variable of interest and not to use `ssh` as a stand-in for water-column thickness.

cbegeman replied:
That's fair. Should we make gh_0 a function of `bottom_depth` then?

xylar replied:
I don't follow.  Isn't it related to water-column thickness, not elevation?

cbegeman replied:
Presumably when users change `bottom_depth` in the config file, `gh_0` should also be changed? But I could be misunderstanding the exact solution in the `_compute_h` function

xylar replied:
No, I don't think that's correct.  `_compute_h` should compute the water-column thickness, which should be independent of the choice of `bottom_depth`.  The `ssh` should be offset by the choice of `bottom_depth` such that `h` (water column thickness) remains the same.

xylar replied:
`h_0 = gh_0 / g` is part of the solution of h, but I don't think it is meant to be a reference elevation but rather a reference water-column thickness.

xylar replied:
But if my understanding isn't correct, it would be very important to know that now, since it could very much affect the error and convergence.

cbegeman replied:
I confirmed that changing the `bottom_depth` does not change the water column thickness, so we don't need to change `gh_0`.

xylar replied:
Okay, great, thank you for checking on that.  It makes me feel a lot more secure about it.

cbegeman reviewed (changes requested):
@xylar Just a few comments while I'm waiting in the queue

xylar commented:
@cbegeman, thanks for catching those mistakes, which are now fixed.  Let me know if there are other changes you would like me to make.

cbegeman reviewed (approved):
I successfully ran both cases without viz; the viz you posted looks good. The code looks good to me, with a few minor changes already noted. Thanks for finishing this case off @xylar and thanks for your contributions during the hackathon @lconlon!

xylar commented:
@cbegeman, I'd prefer to wait until @sbrus89 has time to look at this.  That may very well mean that I will rebase onto #131 once that's done.  That's fine, since either I will have to rebase this or you will have to rebase that and I don't mind the extra work.

sbrus89 commented:
@xylar, I will get to this on Monday. Thanks for your patience.

sbrus89 commented on docs/developers_guide/ocean/tasks/geostrophic.md:
```diff
@@ -0,0 +1,65 @@
+
+# geostrophic
+
+The {py:class}`polaris.ocean.tasks.geostrophic.Geostrophic`
+test performs a series of 24-day runs that advect a bell-shaped tracer blob
+around the sphere.  The resolution of the sphere varies (by default, between
+60 and 240 km).  Advected results are compared with a known exact solution to
+determine the rate of convergence.
```
I believe this description is left over from the `cosine_bell` case

xylar replied:
Oops, thanks!

sbrus89 commented on polaris/ocean/tasks/geostrophic/geostrophic.cfg:
```diff
@@ -0,0 +1,125 @@
+# The type of vertical coordinate (e.g. z-level, z-star)
+coord_type = z-star
+
+# Whether to use "partial" or "full", or "None" to not alter the topography
+partial_cell_type = None
+
+# The minimum fraction of a layer for partial cells
+min_pc_fraction = 0.1
```
Are partial bottom cells something we want to expose in the config file for this case?

xylar replied:
You're right, we could take those out since it wouldn't make sense to use them in this configuration.  I copied them from cosine bell but the same applies there.

sbrus89 commented on polaris/ocean/tasks/geostrophic/geostrophic.cfg:
```diff
@@ -0,0 +1,125 @@
+
+
+# config options for spherical convergence tests
+[spherical_convergence]
```
Do we want to add `icos_resolutions` and `qu_resolutions` here?

xylar replied:
I think it's a good idea to use the default resolutions from:
https://github.com/E3SM-Project/polaris/blob/4bc796e561f8fe1e31c09e34c4abda793d649d6d/polaris/ocean/convergence/spherical/spherical.cfg#L1-L8
That way if we decide later to change the resolutions there, the shared steps in `geostrophic` (and in `cosine_bell`) get updated the same way.  But if you have a reason you think `geostrophic`, `cosine_bell` and others should specify their own, possibly unique list of resolutions, I'm keen to hear it.

sbrus89 reviewed (commented):
Hi @xylar, just a few suggestions as I look through this.

sbrus89 commented:
@xylar - I'm testing a local merge and am seeing this from `polaris list`:

```
Traceback (most recent call last):
  File "/lcrc/group/acme/sbrus/mambaforge/envs/dev_polaris_0.2.0/bin/polaris", line 33, in <module>
    sys.exit(load_entry_point('polaris', 'console_scripts', 'polaris')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lcrc/group/acme/sbrus/mambaforge/envs/dev_polaris_0.2.0/bin/polaris", line 25, in importlib_load_entry_point
    return next(matches).load()
           ^^^^^^^^^^^^^^^^^^^^
  File "/lcrc/group/acme/sbrus/mambaforge/envs/dev_polaris_0.2.0/lib/python3.11/importlib/metadata/__init__.py", line 202, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lcrc/group/acme/sbrus/mambaforge/envs/dev_polaris_0.2.0/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/__main__.py", line 8, in <module>
    from polaris import cache, list, setup, suite
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/list.py", line 6, in <module>
    from polaris.components import get_components
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/components.py", line 5, in <module>
    from polaris.ocean import Ocean
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/ocean/__init__.py", line 4, in <module>
    from polaris.ocean.tasks.geostrophic import add_geostrophic_tasks
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/ocean/tasks/geostrophic/__init__.py", line 6, in <module>
    from polaris.ocean.tasks.geostrophic.analysis import Analysis
  File "/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/ocean/tasks/geostrophic/analysis.py", line 3, in <module>
    from polaris.ocean.convergence.spherical import SphericalConvergenceAnalysis
ImportError: cannot import name 'SphericalConvergenceAnalysis' from 'polaris.ocean.convergence.spherical' (/gpfs/fs1/home/sbrus/polaris_worktrees/williamson/polaris/ocean/convergence/spherical/__init__.py)
```

cbegeman commented:
@xylar Maybe it's worth rebasing onto https://github.com/E3SM-Project/polaris/pull/131 in the process of addressing @sbrus89's issue? Sorry for the hassle.

xylar commented:
@sbrus89, I'm not able to reproduce that issue locally.  Could you try recreating your polaris conda environment to see if that helps?

xylar commented:
@sbrus89, sorry for the confusion on my part yesterday.  I was too tired I guess and I missed that this was with a test merge.  I have rebased and fixed thing up.  Retesting now but I think you can give it another try when you have time.

sbrus89 commented:
@xylar, no problem. Yes I was testing with a local merge with `main`. I'll give it another go today. Thanks for rebasing!

xylar commented:
I successfully tested the `convergence` suite with the `geostrophic` tests added:
```
Task Runtimes:
0:01:01 PASS ocean/planar/inertial_gravity_wave
0:00:55 FAIL ocean/planar/manufactured_solution
0:05:27 PASS ocean/spherical/icos/cosine_bell
0:09:15 PASS ocean/spherical/qu/cosine_bell
0:01:35 PASS ocean/spherical/icos/geostrophic
0:03:02 PASS ocean/spherical/qu/geostrophic
Total runtime: 0:21:16
```

I was having some performance problems with the `geostrphic/with_viz` tests that seem random but also persistent.  I don't know if it's a Chrysalis problem, an issue wiht ESMF or something else.  But I also hope we can replace this viz soon with something better so I'm not inclined to investigate too much.

sbrus89 commented:
@xylar, I'm seeing this error:

```
ocean/spherical/icos/geostrophic
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/sbrus/mambaforge/envs/dev_polaris_0.2.0/bin/polaris", line 33, in <module>
    sys.exit(load_entry_point('polaris', 'console_scripts', 'polaris')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/polaris_worktrees/williamson/polaris/__main__.py", line 62, in main
    commands[args.command]()
  File "/global/cfs/cdirs/e3sm/sbrus/polaris_worktrees/williamson/polaris/run/serial.py", line 196, in main
    run_tasks(suite_name='task', quiet=args.quiet, is_task=True,
  File "/global/cfs/cdirs/e3sm/sbrus/polaris_worktrees/williamson/polaris/run/serial.py", line 103, in run_tasks
    result_str, success, task_time = _log_and_run_task(
                                     ^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/polaris_worktrees/williamson/polaris/run/serial.py", line 316, in _log_and_run_task
    task.steps_to_run = _update_steps_to_run(
                        ^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/polaris_worktrees/williamson/polaris/run/serial.py", line 222, in _update_steps_to_run
    step_str = config.get(task_name, 'steps_to_run').replace(',', ' ')
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/mambaforge/envs/dev_polaris_0.2.0/lib/python3.11/site-packages/mpas_tools/config.py", line 115, in get 
    return self.combined.get(section, option)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/mambaforge/envs/dev_polaris_0.2.0/lib/python3.11/configparser.py", line 797, in get 
    d = self._unify_values(section, vars)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/sbrus/mambaforge/envs/dev_polaris_0.2.0/lib/python3.11/configparser.py", line 1168, in _unify_values
    raise NoSectionError(section) from None
configparser.NoSectionError: No section: 'icos_geostrophic'
```

I looked in the `geostrophic.cfg` file and only see a `icos_geostrophic_with_viz` section. I have a feeling this is because I set up the cases in this order: `ocean/spherical/icos/geostrophic`, `ocean/spherical/icos/geostrophic/with_viz` before running `ocean/spherical/icos/geostrophic`. Is this expected behavior?

sbrus89 commented:
@xylar - I did the following:
```
polaris setup -t ocean/spherical/icos/geostrophic -w $WORKROOT/wiliamson
polaris setup -t ocean/spherical/icos/geostrophic/with_viz -w $WORKROOT/wiliamson 
polaris setup -t ocean/spherical/qu/geostrophic -w $WORKROOT/wiliamson
polaris setup -t ocean/spherical/qu/geostrophic/with_viz -w w $WORKROOT/wiliamson

```
and then 
```
cd $WORKROOT/wiliamson/ocean/spherical/icos/geostrophic
sbatch job_script.sh
```

xylar commented:
@sbrus89, great, thanks!  I'm looking into this now.

Just as a tip to make your workflow more efficient, you may want to do:
```
polaris list
```
and then:
```
polaris setup -n 17 18 19 20 -w $WORKROOT/wiliamson
```
This will set up a custom suite with all 4 tasks.  But it doesn't seem to expose the problem you showed so I'm glad you did things the "more tedious" way.

xylar commented:
@sbrus89, I understand now what is going wrong and I don't have a good idea what to do about it yet.  It isn't anything to do with this PR, it was caused by #125.  When you separately set up 2 tasks that share the same config file, the first task will create a section, option and value with its `steps_to_run`.  Then, the second task will overwrite the config file, including its section, option and value with *its* `steps_to_run`.  If you were to set them up together, both sections, options and values would be populated as excepted:
``` ini
[icos_geostrophic]

# A list of steps to include when running the icos_geostrophic task
# source: /gpfs/fs1/home/ac.xylar/e3sm_work/polaris/add-geostrophic-tests/polaris/setup.py
steps_to_run = icos_base_mesh_60km icos_init_60km icos_forward_60km icos_base_mesh_120km icos_init_120km icos_forward_120km icos_base_mesh_240km icos_init_240km icos_forward_240km icos_base_mesh_480km icos_init_480km icos_forward_480km analysis


[icos_geostrophic_with_viz]

# A list of steps to include when running the icos_geostrophic_with_viz task
# source: /gpfs/fs1/home/ac.xylar/e3sm_work/polaris/add-geostrophic-tests/polaris/setup.py
steps_to_run = icos_base_mesh_60km icos_init_60km icos_forward_60km icos_map_cell_60km icos_map_edge_60km icos_viz_60km icos_base_mesh_120km icos_init_120km icos_forward_120km icos_map_cell_120km icos_map_edge_120km icos_viz_120km icos_base_mesh_240km icos_init_240km icos_forward_240km icos_map_cell_240km icos_map_edge_240km icos_viz_240km icos_base_mesh_480km icos_init_480km icos_forward_480km icos_map_cell_480km icos_map_edge_480km icos_viz_480km analysis
```

Presumably, the solution will be to include `steps_to_run` from all tasks when writing out a shared config file.  Since this is unrelated to `geostrophic`, I'll address it in a separate PR.

xylar commented:
@sbrus89, would you like to review #134 and for me to merge that and rebase? Or can you review this without that fix?

sbrus89 commented:
@xylar, I figured it wasn't related to this PR. I'll just review it without the fix. I'll try it by creating a custom suite as you suggested. 

Just a thought, it would be nice to be able to create custom suites with the test case names as opposed to numbers since those change as new tests are added.

xylar commented:
@sbrus89, that's a good idea and easy to implement.  I'll make a PR shortly.

xylar commented:
@sbrus89, hmm, that looks like a bug where a relative path didn't get turned into an absolute path as it should have before it got stored in the pickle file.  I have never seen that before, so I would need to know exactly what your `polaris setup` or `polaris suite` command was that led to this.  Maybe it was a simple as not supplying `-p`, which I basically always do?

sbrus89 commented:
Sounds like it could be from not supplying `-p` I don't typically do this unless I want an `ocean_model` outside the repo.

xylar commented:
@sbrus89, I'm not able to reproduce this with `main` even without the `-p` flag.  I doubt I introduced this issue in this PR.  Could you give me the exact `polaris setup` or `polaris suite` command?

xylar commented:
Oh, I take that back!  I see it only in some tests and not others:
```
0:00:08 PASS ocean/planar/baroclinic_channel/10km/threads
0:00:03 PASS ocean/planar/baroclinic_channel/10km/decomp
0:00:00 FAIL ocean/planar/baroclinic_channel/10km/restart
0:00:01 FAIL ocean/planar/inertial_gravity_wave
0:00:02 PASS ocean/single_column/cvmix
0:00:02 PASS ocean/single_column/ideal_age
```

Again, not related to this PR but an important bug to fix.

xylar commented:
@sbrus89, I see what changed and why.  It's complicated to explain (to do with fancy config parsing and automatically converting to absolute paths) but probably easy to fix.  Thanks for bringing this to my attention.  Can you proceed with using the `-p` flag for now?

sbrus89 commented:
@xylar - yes, will do. Sorry for the headache on this!

sbrus89 reviewed (approved):
This looks great to me. Thanks for your work on this and for addressing these mostly unrelated issues @xylar!

xylar commented:
Thanks so much, @sbrus89!  And I really do appreciate you finding those issues, whether they are related or not.

xylar commented:
@cbegeman, thank you very much for getting this test case started with excellent documentation and for thorough review!  @lconlon, thank you for helping me get this test case started during the hackathon!
